core/node/node_framework/src/implementations/layers/web3_api/tx_sender.rs

```diff
@@ -0,0 +1,160 @@
+use std::{fmt, sync::Arc};
+
+use zksync_core::api_server::{
+    execution_sandbox::{VmConcurrencyBarrier, VmConcurrencyLimiter},
+    tx_sender::{ApiContracts, TxSenderBuilder, TxSenderConfig},
+};
+use zksync_state::PostgresStorageCaches;
+
+use crate::{
+    implementations::resources::{
+        fee_input::FeeInputResource,
+        pools::ReplicaPoolResource,
+        state_keeper::ConditionalSealerResource,
+        web3_api::{TxSenderResource, TxSinkResource},
+    },
+    service::{ServiceContext, StopReceiver},
+    task::Task,
+    wiring_layer::{WiringError, WiringLayer},
+};
+
+#[derive(Debug)]
+pub struct PostgresStorageCachesConfig {
+    pub factory_deps_cache_size: u64,
+    pub initial_writes_cache_size: u64,
+    pub latest_values_cache_size: u64,
+}
+
+#[derive(Debug)]
+pub struct TxSenderLayer {
+    tx_sender_config: TxSenderConfig,
+    postgres_storage_caches_config: PostgresStorageCachesConfig,
+    max_vm_concurrency: usize,
+    api_contracts: ApiContracts,
+}
+
+impl TxSenderLayer {
+    pub fn new(
+        tx_sender_config: TxSenderConfig,
+        postgres_storage_caches_config: PostgresStorageCachesConfig,
+        max_vm_concurrency: usize,
+        api_contracts: ApiContracts,
+    ) -> Self {
+        Self {
+            tx_sender_config,
+            postgres_storage_caches_config,
+            max_vm_concurrency,
+            api_contracts,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl WiringLayer for TxSenderLayer {
+    fn layer_name(&self) -> &'static str {
+        "tx_sender_layer"
+    }
+
+    async fn wire(self: Box<Self>, mut context: ServiceContext<'_>) -> Result<(), WiringError> {
+        // Get required resources.
+        let tx_sink = context.get_resource::<TxSinkResource>().await?.0;
+        let pool_resource = context.get_resource::<ReplicaPoolResource>().await?;
+        let replica_pool = pool_resource.get().await?;
+        let sealer = match context.get_resource::<ConditionalSealerResource>().await {
```

Comment:
Future work (?): May make sense to add `get_optional_resource` method that would return `Result<Option<T>, _>`; seems like a somewhat common pattern.

Reply:
Yeah, I'm also thinking about this. Not sure about it just yet, since such optional resources themselves feel a bit wrong -- maybe the correct approach would be to know for sure whether you need the resource or not. Otherwise, we can create a surface for implicit bugs where some component is missing by mistake, but the task still runs without it, even if that's not what we want.